proposed/flexible-http.md

```diff
@@ -0,0 +1,149 @@
+
+- For discussion sake (actual API TBD), we will consider that this API consists of an abstraction and concrete implementations:
+    - `HttpConnection`, the abstraction.
+    - `Http1Connection`, `Http2Connection`, `Http3Connection`, the low-level implementations.
+
+## Low-level API
+
+Low-level APIs implement HTTP without any additional features or opinions. They return HTTP data direct to the user without transformation, and are built to be forward-compatible with new HTTP features (e.g. HTTP/2 and HTTP/3 extension frames) in the future.
+
+Performance is of high priority:
+
+- Operate on `Span<byte>` rather than `string` and `Uri`.
+    - Parameters will not be validated (to enable efficient forwarding), but helpers will be exposed to allow user to perform validation.
+- Zero allocations (steady-state).
+
+Ease of use is an explicit non-goal:
+
+- We assume users will know more about HTTP than `HttpClient` currently requires of them.
+- It is okay if it takes more code to use than `HttpClient`.
+
+These APIs are constructed on top of `Stream`. This has some implications worth noting:
+
+- Connections are not established by these APIs: the user must establish the connection.
+- Connecting via a proxy, or over TLS is not done by this API: the user must pass in a `Stream` that already has these established.
+
+### Header handling
+
+Headers are read and written "streaming", without going through a collection.
+
+"Prepared" headers will enable dynamic table compression and pre-validation/serialization equivalent to existing internal "known header" concept of `SocketsHttpHandler`.
+
+Huffman compression will be opt-in on a per-header basis.
+
+### Content handling
+
+Content will be read and written directly to the `HttpConnection`'s request type, without using `Stream` or a `HttpContent`-equivalent, to avoid the associated allocations.
+
+Expect Continue will not be implemented here, but full informational responses will be returned to allow the caller to implement their own Expect Continue processing.
+
+## Compositional APIs
+
+Compositional APIs are built on top of low-level APIs to provide additional features, as well as some opinionated implemenations.
+
+Every higher-level feature that `SocketsHttpHandler` currently supports should have an implementation here. This will mean one or more `HttpConnection` that implements some features:
+
+- Automatic decompression.
+- Automatic redirects.
+- Connection pooling.
+- Version negotiation.
+    - ALPN, Upgrade header, ALT-SVC.
+- Authentication.
+
+As well as some additional APIs for working with the `HttpConnection` API:
+
+- Header parsing/printing.
+- Input validation.
+- Proxy resolver.
+    - "HTTP_PROXY" etc. environment variables.
+    - WinInet settings.
+- Proxy implemenations.
+    - HTTP proxy as `HttpConnection`.
+    - CONNECT proxy as `Stream`.
+    - SOCKS proxy as `Stream`.
+- Wrapper to work with `HttpConnection` as a `Stream`.
```

Comment:
Somewhere in here, we should call out support for WebSockets, specifically the Upgrade: websocket scenario.

Reply:
Are you thinking a full WebSocket API, or just the ability to "dispose and extract unconsumed read buffer" from a connection?

Reply:
Mostly the latter. IIRC, WebSockets require CONNECT tunneling and Upgrade: support to get the raw stream. We just need to ensure that those are supported at this level. WebSocket functionality can be built on top.